Further changes to target-independent files from Morello PR

arichardson opened the pull request.

LawrenceEsswood committed:
Add cap load helper that returns un-squashed tag

Needed to implement capability atomics on Morello.

LawrenceEsswood committed:
Add tag memory property to automatically create tags for memory objects

LawrenceEsswood committed:
Factor out mmu_index from tag set/get

Needed for Morello.

LawrenceEsswood committed:
Fix segfault in debug message

LawrenceEsswood committed:
Fix argument in cheri_tag_set_many

LawrenceEsswood committed:
[cheri-compressed-cap] Fix Morello bounds setting sign extension

LawrenceEsswood committed:
[Morello] Check capability decode is valid in set bounds

LawrenceEsswood committed:
[cheri-compressed-cap] Cache the exponent value

Storing this in a dedicated field will allow implementing the fast
representability check in TCG.

arichardson committed:
Remove unused pcc_check_load helper

arichardson committed:
Remove calls to cheri_tag_init that are no longer needed

jrtc27 commented on backends/hostmem.c:
```diff
@@ -320,6 +320,13 @@ size_t host_memory_backend_pagesize(HostMemoryBackend *memdev)
 #endif
 
+void cheri_tag_init(MemoryRegion *mr, uint64_t memory_size);
+__attribute__((weak)) void cheri_tag_init(MemoryRegion *mr,
```
Blegh this is all gross but I guess is because backends is built once for all targets so you can't ifdef based on TARGET_FOO?

LawrenceEsswood replied:
Yeah =/

arichardson replied:
Yeah this is a bit ugly, but seems like the easiest workaround for TARGET_FOO.

jrtc27 commented on target/cheri-common/cheri-compressed-cap/cheri_compressed_cap_common.h:
```diff
@@ -838,8 +840,13 @@ static inline bool _cc_N(setbounds_impl)(_cc_cap_t* cap, _cc_addr_t req_base, _c
     _cc_length_t new_top = new_cap._cr_top;
 
     if (exact) {
+#ifndef CC_IS_MORELLO
```
This diff is unrelated to the commit it's in (and is Morello-specific so shouldn't be in this PR?)

arichardson replied:
I meant "touches target-independent files" in the PR title, but I agree we should split commits.

arichardson replied:
Merged into the previous commit.

jrtc27 commented on tcg/tcg-op.c:
```diff
@@ -2829,9 +2829,25 @@ static inline void plugin_gen_mem_callbacks(TCGv_cap_checked_ptr vaddr, uint16_t
 #endif
 }
 
+#ifdef TARGET_AARCH64
```
Not the way to do it, and doesn't belong in this PR

jrtc27 commented on tcg/tcg-op.c:
```diff
@@ -2829,9 +2829,25 @@ static inline void plugin_gen_mem_callbacks(TCGv_cap_checked_ptr vaddr, uint16_t
+     * If idx always matches whatever would be used by calling cpu_mmu_index(),
+     * there is no point of including the extra argument on this hot path.
+     */
     gen_helper_cheri_invalidate_tags(cpu_env, out_addr, memop);
```
Should it be an error to ever use this on AArch64? If so, perhaps it shouldn't be available there?

LawrenceEsswood replied:
To use what sorry? gen_helper_cheri_invalidate_tags? No, it is fair any case where tcg_idx was what cpu_mmu_index would return. Just no (obvious) way of knowing in target-general way at translation time that this is the case, unless it is always true (which I think it is for RISCV and MIPS). The only time I know this actual happens is the family of arm instructions that make accesses as if they were made at EL0.

arichardson replied:
I'd have to check, but in some cases it might still make sense to call the helper without the explicit argument for Morello since it avoids the extra tcg temp. CC: @LawrenceEsswood

LawrenceEsswood replied:
It would, but then then we would either need to pass a flag that indicates an access is to a default mmu_idx (ugly, messy refactor), or somehow have a default mmu_idx stored in the translation context to compare to. Arm already has a default mmu_idx in it's DisasContext, but DisasContextBase (which is what we have access to here) does not. I don't think we want to either. We could, of course, add a 'is_default_mmu_idx' to a number of target specific headers (which can get DisasContext with a bit of container of magic), and have it return false for targets that do not have such a notion, or true if its always true. Seems like a mess to avoid an extra constant load in tcg.

arichardson replied:
Did you measure how much it slows down non-AArch64? Maybe we should just always pass it?

arichardson replied:
Fixed in latest version

arichardson committed:
Always pass mmu_idx to cheri_invalidate_tags

We can use TCGMemOpIdx to pass both MemOp and mmu_idx to the helper without
requiring an additional TCG temp. This allows us to drop the AARCH64 ifdef
here and always respect the explicit mmu_idx even for RISC-V/MIPS.

arichardson commented on tcg/tcg-op.c:
```diff
@@ -2960,19 +2938,19 @@ void tcg_gen_qemu_st_i32_with_checked_addr(TCGv_i32 val, TCGv_cap_checked_ptr ad
 
     plugin_gen_mem_callbacks(addr, info);
 #if defined(TARGET_MIPS) || defined(TARGET_RISCV) || defined(CONFIG_TCG_LOG_INSTR)
```
hmm this should probably be `defined(TARGET_CHERI)`

LawrenceEsswood replied:
gen_cheri_break_loadlink needs a tcop as well, even if not CHERI,

arichardson replied:
yeah just noticed that, so we need `defined(TARGET_MIPS) || defined(TARGET_RISCV) || defined(TARGET_CHERI)`

arichardson replied:
Removed the extra arg.

arichardson committed:
Drop memop argument from gen_cheri_break_loadlink()

It's unused and removing it allows us to simplify #ifdefs in tcg-op.c

jrtc27 commented on include/exec/helper-head.h:
```diff
@@ -48,6 +49,8 @@
 #define dh_ctype_void void
 #define dh_ctype_memop MemOp
+/* Can't use TCGMemOpIdx here due to include ordering. */
+#define dh_ctype_memop_idx uint32_t /* TCGMemOpIdx */
```
Did you actually verify that? Because this is a #define, TCGMemOpIdx only needs to be defined when this macro is used, not here.

arichardson replied:
```[2/97] Compiling C object libqemu-riscv64cheri-softmmu.fa.p/accel_tcg_tcg-runtime-gvec.c.o
In file included from ../../qemu/accel/tcg/tcg-runtime-gvec.c:23:
In file included from /Users/alex/cheri/qemu/include/exec/helper-proto.h:40:
/Users/alex/cheri/qemu/accel/tcg/tcg-runtime.h:154:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
DEF_HELPER_3(cheri_invalidate_tags, void, env, cap_checked_ptr, memop_idx)
^
/Users/alex/cheri/qemu/include/exec/helper-head.h:163:5: note: expanded from macro 'DEF_HELPER_3'
    DEF_HELPER_FLAGS_3(name, 0, ret, t1, t2, t3)
    ^
/Users/alex/cheri/qemu/include/exec/helper-proto.h:19:57: note: expanded from macro 'DEF_HELPER_FLAGS_3'
dh_ctype(ret) HELPER(name) (dh_ctype(t1), dh_ctype(t2), dh_ctype(t3));
                                                        ^
/Users/alex/cheri/qemu/include/exec/helper-head.h:55:21: note: expanded from macro 'dh_ctype'
#define dh_ctype(t) dh_ctype_##t
                    ^
<scratch space>:217:1: note: expanded from here
dh_ctype_memop_idx
^
/Users/alex/cheri/qemu/include/exec/helper-head.h:53:28: note: expanded from macro 'dh_ctype_memop_idx'
#define dh_ctype_memop_idx TCGMemOpIdx
                           ^
```

arichardson replied:
I guess we could move those types to exec/memop.h but I'd prefer if we keep upstream diffs small.

jrtc27 replied:
Yeah ok just tried it locally and there are a lot of those cases
